### PR TITLE
Implement BananaButtonStateManager

### PR DESCRIPTION
**Pull request by jdragonbae**
This patch implements BananaButtonStateManager for updating the button
state. Via hooking the upstream code, it will notify listeners of the
button state.

Refs: #1003

### CODE REVIEW

File: src/org/triple/banana/toolbar/ButtonStateDispatcher.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 The Triple Banana Authors. All rights reserved.
```
Comment (romandev) [inline src/org/triple/banana/toolbar/ButtonStateDispatcher.java:1]: Do you have any reason to put this code into `toolbar` directory?

**Review by romandev — COMMENTED** [review]
IMHO, you might have to draw your idea as a class diagram in order to describe your idea to us.

**Comment by jdragonbae** [comment]
> IMHO, you might have to draw your idea as a class diagram in order to describe your idea to us.

Thanks for the suggestion. I attached a class diagram on the issue[1].

[1] https://github.com/triplebanana/triple_banana/issues/1003#issuecomment-757439790

File: src/org/triple/banana/toolbar/ButtonStateDispatcher.java (hunk 1 above)
Comment (jdragonbae) [reply src/org/triple/banana/toolbar/ButtonStateDispatcher.java:1]: There is no special reason. I can put this code into another directory like `button_state`.

File: src/org/triple/banana/button_state/ButtonStateDispatcher.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import org.triple.banana.R;
+import org.triple.banana.theme.DarkModeController;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class ButtonStateDispatcher implements BananaButtonStateDispatcher {
+    interface Listener {
+        void onUpdateButtonState(
+                @IdRes int id, @DrawableRes int image, @ColorRes int color, boolean isEnabled);
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateDispatcher.java:24]: Do we need the `image` and `color`? The event should be fired with the button state. Then, the client can change the values such as `image`, `color`, can't it?

File: src/org/triple/banana/button_state/ButtonStateDispatcher.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import androidx.annotation.Nullable;
+
+import org.banana.cake.interfaces.BananaButtonStateDispatcher;
+import org.triple.banana.R;
+import org.triple.banana.theme.DarkModeController;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class ButtonStateDispatcher implements BananaButtonStateDispatcher {
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateDispatcher.java:21]: ButtonStateManager might be better because the class's role is two things. One thing is event routing to client and another is event listening from upstream.

File: src/org/triple/banana/button_state/ButtonStateDispatcher.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
+// Copyright 2021 The Triple Banana Authors. All rights reserved.
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateDispatcher.java:1]: How can we get the current state? If the button info data is loaded in ViewModel, then we should set the initial button state but this implementation can only listen to the state change event if the button state is actually changed.

File: src/org/triple/banana/button_state/ButtonStateDispatcher.java (hunk 3 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateDispatcher.java:21]: Ack. I'll change the name of this class.

File: src/org/triple/banana/button_state/ButtonStateDispatcher.java (hunk 4 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateDispatcher.java:1]: Thanks for the review. You're right. For getting current state, I'll try to cache each button states. And then notifies this values when add listener.

File: src/org/triple/banana/button_state/ButtonStateDispatcher.java (hunk 2 above)
Comment (romandev) [reply src/org/triple/banana/button_state/ButtonStateDispatcher.java]: @jdragonbae You missed my comment here.

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        void onUpdateButtonState(
+                @IdRes int id, @DrawableRes int image, @ColorRes int color, boolean isEnabled);
+    }
+
+    static class CachedButtonState {
+        boolean isBookmarked;
+        boolean editingAllowed;
+        boolean canGoBack;
+        boolean canGoForward;
+        boolean isDesktopView;
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:32]: I think the button state should be `enum` instead of `boolean`.
Also, this button state should be cahced by HashMap.

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import org.triple.banana.R;
+import org.triple.banana.theme.DarkModeController;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class ButtonStateManager implements BananaButtonStateManager {
+    interface Listener {
+        void onUpdateButtonState(
+                @IdRes int id, @DrawableRes int image, @ColorRes int color, boolean isEnabled);
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:24]: If the button state would be `enum` instead of `boolean`, you could represent various state.

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        }
+        return sInstance;
+    }
+
+    public void addListener(@NonNull Listener listener) {
+        mListeners.add(listener);
+        updateBookmarkButtonStatus(mCachedState.isBookmarked, mCachedState.editingAllowed);
+        updateBackButtonVisibility(mCachedState.canGoBack);
+        updateForwardButtonVisibility(mCachedState.canGoForward);
+        updateDesktopViewButtonState(mCachedState.isDesktopView);
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:53]: This is something wrong. Why do you want to re-assign the existing cache values?

File: src/org/triple/banana/button_state/ButtonStateDispatcher.java (hunk 2 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateDispatcher.java]: The reason why the image and color parameters are added is for bookmark button status. To decide the color and image of the bookmark, need two parameters (isBookmarked, isEditingAllowed). So, I thought `onUpdateButtonState` is called with the image and color decided.

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            sInstance = new ButtonStateManager();
+        }
+        return sInstance;
+    }
+
+    public void addListener(@NonNull Listener listener) {
+        mListeners.add(listener);
+        updateBookmarkButtonStatus(mCachedState.isBookmarked, mCachedState.editingAllowed);
+        updateBackButtonVisibility(mCachedState.canGoBack);
+        updateForwardButtonVisibility(mCachedState.canGoForward);
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:52]: BTW, this looks that these buttons are not in quick menu candidate list, am I missing something?

Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:52]: Also, I think you should have considered `Download`, `Find in page`, `Add to home`, `Share` and `Print` instead?

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 6 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java:24]: Thank you. It helped me a lot.

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 5 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java:32]: Ack.

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 7 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java:53]: Ack.

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 8 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java:52]: I think the bottom toolbar would be another client. So, I added.

Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java:52]: As you said, we need consider `Download`, `Find in page`, `Add to home`, `Share` and `Print`.
I'll upload about that.

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    public void updateForwardButtonVisibility(boolean canGoForward) {
+        final ForwardStatus status = canGoForward ? ForwardStatus.ENABLED : ForwardStatus.DISABLED;
+        notifyListeners(R.id.forward, status);
+    }
+
+    @Override
+    public void updateDesktopViewButtonState(boolean isDesktopView) {
+        final DesktopViewStatus status =
+                isDesktopView ? DesktopViewStatus.DESKTOP_VIEW : DesktopViewStatus.MOBILE_VIEW;
+        notifyListeners(R.id.desktop_view, status);
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:100]:
```
public void updateDesktopPageButtonState(boolean isDesktopPage) {
    ButtonState state = isDesktopPage ? ButtonState.DESKTOP_PAGE : ButtonState.MOBILE_PAGE;
    mCachedState.put(R.id.desktop_page, state);
    notifyListeners(R.id.desktop_page, state);
}
```

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    private void notifyListeners(@IdRes int id, @NonNull Enum<?> status) {
+        mCachedStatus.put(id, status);
+        for (final Listener listener : mListeners) {
+            listener.onUpdateButtonState(id, status);
+        }
+    }
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:108]:
```
private void notifyListeners(@IdRes int id, @NonNull ButtonState state) {
    for (Listener listener : mListeners) {
        listener.onUpdateButtonState(id, state);
    }
}
```

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        BOOKMARKED,
+        NOT_BOOKMARKED,
+        DISABLED,
+    }
+
+    public enum DesktopViewStatus {
+        DESKTOP_VIEW,
+        MOBILE_VIEW,
+        DISABLED,
+    }
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Thanks for the review. Done.

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+    private static @Nullable ButtonStateManager sInstance;
+    private final @NonNull Set<Listener> mListeners = new HashSet<>();
+    private final @NonNull Map<Integer, Enum<?>> mCachedStatus = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Done

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Override
+    public void updateBookmarkButtonStatus(boolean isBookmarked, boolean editingAllowed) {
+        BookmarkStatus status = BookmarkStatus.BOOKMARKED;
+        if (!editingAllowed) {
+            status = BookmarkStatus.DISABLED;
+        } else if (!isBookmarked) {
+            status = BookmarkStatus.NOT_BOOKMARKED;
+        }
+        notifyListeners(R.id.bookmark, status);
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Done

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    @Override
+    public void updateBackButtonVisibility(boolean canGoBack) {
+        final BackStatus status = canGoBack ? BackStatus.ENABLED : BackStatus.DISABLED;
+        notifyListeners(R.id.back, status);
+    }
+
+    @Override
+    public void updateForwardButtonVisibility(boolean canGoForward) {
+        final ForwardStatus status = canGoForward ? ForwardStatus.ENABLED : ForwardStatus.DISABLED;
+        notifyListeners(R.id.forward, status);
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Done

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 9 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Done

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 10 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Done

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class ButtonStateManager implements BananaButtonStateManager {
+    public interface Listener {
+        void onUpdateButtonState(@IdRes int id, @NonNull Enum<?> state);
+    }
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Done

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+        } else if (!isBookmarked) {
+            status = BookmarkStatus.NOT_BOOKMARKED;
+        }
+        notifyListeners(R.id.bookmark, status);
+    }
+
+    @Override
+    public void updateBackButtonVisibility(boolean canGoBack) {
+        final BackStatus status = canGoBack ? BackStatus.ENABLED : BackStatus.DISABLED;
+        notifyListeners(R.id.back, status);
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Done

**Comment by jdragonbae** [comment]
@romandev Thanks for your time to review. Would it be okay to implement the client in a new patch?

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    public void removeListener(@NonNull Listener listener) {
+        if (mListeners.contains(listener)) {
+            mListeners.remove(listener);
+        }
+    }
+
+    @Override
+    public void updateBookmarkButtonStatus(boolean isBookmarked, boolean editingAllowed) {
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:61]: You didn't update this line yet.

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+            state = ButtonState.BOOKMARKED;
+        } else {
+            state = ButtonState.NOT_BOOKMARKED;
+        }
+        mCachedState.put(R.id.bookmark, state);
+        notifyListeners(R.id.bookmark, state);
+    }
+
+    @Override
+    public void updateBackButtonVisibility(boolean canGoBack) {
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:75]: You didn't update this line yet.

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+        final ButtonState state = canGoBack ? ButtonState.ENABLED : ButtonState.DISABLED;
+        mCachedState.put(R.id.back, state);
+        notifyListeners(R.id.back, state);
+    }
+
+    @Override
+    public void updateForwardButtonVisibility(boolean canGoForward) {
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:82]: You didn't update this line yet.

File: src/org/triple/banana/button_state/ButtonStateManager.java [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+        final ButtonState state = canGoForward ? ButtonState.ENABLED : ButtonState.DISABLED;
+        mCachedState.put(R.id.forward, state);
+        notifyListeners(R.id.forward, state);
+    }
+
+    @Override
+    public void updateDesktopViewButtonState(boolean isDesktopView) {
```
Comment (romandev) [inline src/org/triple/banana/button_state/ButtonStateManager.java:89]: You didn't update this line yet.

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 17 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Sorry, I missed it.

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 18 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Updated to `updateBackButtonState`

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 19 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Updated to `updateForwardButtonState`

File: src/org/triple/banana/button_state/ButtonStateManager.java (hunk 20 above)
Comment (jdragonbae) [reply src/org/triple/banana/button_state/ButtonStateManager.java]: Updated to `updateDesktopPageButtonState`

**Comment by romandev** [comment]
> Thanks for your time to review. Would it be okay to implement the client in a new patch?

IMHO, it's better to add the client part in this patch as well.
It is not large and we should check whether this code is actually working or not.

**Comment by jdragonbae** [comment]
> > Thanks for your time to review. Would it be okay to implement the client in a new patch?
> 
> IMHO, it's better to add the client part in this patch as well.
> It is not large and we should check whether this code is actually working or not.

Thanks! I'll upload with the client part.

File: src/org/triple/banana/quick_menu/ButtonInfo.java [hunk 21]
```diff
@@ -12,8 +12,10 @@
 
 class ButtonInfo implements Model.Data {
     final @IdRes int id;
-    final @IdRes int image;
-    final @StringRes int label;
+    @IdRes
+    int image;
+    @StringRes
+    int label;
```
Comment (romandev) [inline src/org/triple/banana/quick_menu/ButtonInfo.java:18]: This is unnecessary change

File: src/org/triple/banana/quick_menu/ViewControllerImpl.java [hunk 22]
```diff
@@ -8,27 +8,58 @@
 import androidx.annotation.IdRes;
 import androidx.annotation.NonNull;
 
+import org.triple.banana.R;
+import org.triple.banana.button_state.ButtonStateManager;
+import org.triple.banana.button_state.ButtonStateManager.ButtonState;
+import org.triple.banana.button_state.ButtonStateManager.Listener;
+
 class ViewControllerImpl implements ViewController {
```
Comment (romandev) [inline src/org/triple/banana/quick_menu/ViewControllerImpl.java:16]: implements ViewController, ButtonStateManager.Listener

File: src/org/triple/banana/quick_menu/ViewModelImpl.java [hunk 23]
```diff
@@ -5,23 +5,34 @@
...
 
     @Override
     public @NonNull List<ButtonInfo> getButtons() {
         return mButtons;
     }
+
+    @Nullable
+    ButtonInfo getButtonById(@IdRes int id) {
+        return mButtonMap.get(id);
+    }
```
Comment (romandev) [inline src/org/triple/banana/quick_menu/ViewModelImpl.java:37]:
```
void updateButton(ButtonInfo buttonInfo) {
    mButtonMap.put(buttonInfo.id, buttonInfo);
}
```

File: src/org/triple/banana/quick_menu/ViewControllerImpl.java [hunk 24]
```diff
@@ -8,27 +8,58 @@
...
+            if (id == R.id.desktop_view) {
+                ButtonInfo info = mViewModel.getButtonById(id);
+                if (info == null) {
+                    return;
+                }
+                info.image = state == ButtonState.MOBILE_PAGE ? R.drawable.ic_phone_black_24dp
+                                                              : R.drawable.ic_desktop_black_24dp;
+                info.label = state == ButtonState.MOBILE_PAGE ? R.string.mobile_view
+                                                              : R.string.desktop_view;
+                mViewModel.notifyViews();
```
Comment (romandev) [inline src/org/triple/banana/quick_menu/ViewControllerImpl.java:61]:
```
if (ButtonState.MOBILE_PAGE) {
    image = ...;
    label = ...;
} else {
    image = ...;
    label = ...;
}
mViewModel.updateButton(new ButtonInfo(id, image, label));
mViewModel.notifyViews();
```

File: src/org/triple/banana/quick_menu/ViewModelImpl.java (hunk 23 above)
Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/ViewModelImpl.java:37]: Done

File: src/org/triple/banana/quick_menu/ViewControllerImpl.java (hunk 24 above)
Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/ViewControllerImpl.java:61]: Done.

File: src/org/triple/banana/quick_menu/ViewControllerImpl.java (hunk 22 above)
Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/ViewControllerImpl.java:16]: Done

File: src/org/triple/banana/quick_menu/ButtonInfo.java (hunk 21 above)
Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/ButtonInfo.java:18]: Thanks for the review. Done.

File: src/org/triple/banana/quick_menu/ViewModelImpl.java [hunk 25]
```diff
@@ -6,22 +6,29 @@
...
     }
 
     @Override
     public @NonNull List<ButtonInfo> getButtons() {
-        return mButtons;
+        return new ArrayList<ButtonInfo>(mButtons.values());
+    }
+
+    void updateButton(@NonNull ButtonInfo buttonInfo) {
+        mButtons.put(buttonInfo.id, buttonInfo);
```
Comment (romandev) [inline src/org/triple/banana/quick_menu/ViewModelImpl.java:32]: This code might create a new key-value pair if there was no existing key-value in the map. So, it's better to add `contains` check to avoid the situation.

Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/ViewModelImpl.java:32]: Ack

File: src/org/triple/banana/quick_menu/ViewModelImpl.java [hunk 26]
```diff
@@ -5,23 +5,38 @@
...
+            mButtonMap.put(buttonInfo.id, buttonInfo);
         }
     }
 
     @Override
     public @NonNull List<ButtonInfo> getButtons() {
         return mButtons;
     }
+
+    void updateButton(@IdRes int id, @IdRes int image, @StringRes int label) {
```
Comment (romandev) [inline src/org/triple/banana/quick_menu/ViewModelImpl.java:34]:
```
void updateButton(@NonNull ButtonInfo newInfo) {
    ButtonInfo existingInfo = mButtonMap.get(newInfo.id);
    if (existingInfo == null) return;

    existingInfo.image = newInfo.image;
    existingInfo.label = newInfo.label;
}
```

Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/ViewModelImpl.java]: Done

**Review by romandev — APPROVED** [review]
lgtm
